### ISSUE TITLE
Cloning and Emulating a Jailbroken iPod Touch 1G

### ISSUE THREAD

**Issue by FrederickGeek8** [devos50/qemu#31]
Hi there,

The work you've been doing is really amazing! I have a working jailbroken iPod Touch 1G that is dear to me and I was hoping I could eventually preserve. 

Do you know how I could go about dumping everything from the device I'd need to clone it the best I can?

Thanks!

**Comment by devos50** [comment]
I appreciate your interest in this work! I would say that the most important thing to dump is the filesystem. While I don't provide any tools for that, you might find work done around the release of the iPod Touch 1G or 2G that would allow such functionality. Perhaps [openiboot](https://github.com/iDroid-Project/openiBoot) would also support this?

**Comment by zoe-vb** [comment]
You could probably use dd to backup the device nand and then set the output file to be sent over scp to another machine

**Comment by FrederickGeek8** [comment]
I was having troubles keeping an SSH (over iproxy) connection open long enough to a `tar` of the root. I'm not sure if it's a problem with the OS/software but I ended up using `rsync` to pull everything over in a recoverable way. I will pick up an official 30-pin connector (instead of a cheap Amazon one) and hopefully that will allow me to not drop my SSH connection and use `dd` or friends.

I tried `rsync`'ing `/` to a folder on my Mac and created a HFS `dmg` file from that folder. I then ran the `dmg2img` and generated my NAND succesfully. The emulator won't boot though. The serial output is non-descriptive.

@devos50 is there any special magic in the `filesystem-writeable.dmg` that would differ from a stock image? It's possible that my `rsync`-based procedures just botched things. I'm also worried that maybe there is some difference in the system that might not be accounted for -- like how my device has a custom Pineapple boot logo stock from the jailbreak (I think it was an early MuscleNerd production).

EDIT: It turns out that this iPod Touch 1G is running iPhone OS 2.2.1. Oops. Is this still feasible?

Serial output after iBoot init:
```
Reading PMU register 87
[FTL:MSG] Apple NAND Driver (AND) 0x43303032
[NAND] Device ID           0xa514d3ad
[NAND] BANKS_TOTAL         8
[NAND] BLOCKS_PER_BANK     4096
[NAND] SUBLKS_TOTAL        4096
[NAND] USER_SUBLKS_TOTAL   3872
[NAND] PAGES_PER_SUBLK     1024
[NAND] PAGES_PER_BANK      524288
[NAND] SECTORS_PER_PAGE    4
[NAND] BYTES_PER_SPARE     64
[FTL:MSG] FIL_Init			[OK]
[FTL:MSG] BUF_Init			[OK]
[FTL:MSG] VFL_Init			[OK]
[FTL:MSG] FTL_Init			[OK]
[FTL:MSG] VFL_Open			[OK]
[FTL:MSG] FTL_Open			[OK]
Reading PMU register 105
Boot Failure Count: 0	Panic Fail Count: 0
Delaying boot for 0 seconds. Hit enter to break into the command prompt...
HFSInitPartition: 0x1802b8f0
Reading PMU register 75
power supply type batt
battery voltage Reading PMU register 87
error
USBMPInit()
Endpoint direction info:
EP0:	BIDIR
EP1:	BIDIR
EP2:	BIDIR
EP3:	BIDIR
EP4:	BIDIR
EP5:	BIDIR
usbfc_menu_init()
Entering recovery mode, starting command prompt
] Reading PMU register 75
```

**Comment by zoe-vb** [comment]
The fstab needs changing, also I believe the kernel was also moved. But are the permissions of the file system correct?

**Comment by devos50** [comment]
@zoe-vb I indeed changed the `fstab` file so there's just one partition. I merged the `boot` partition with the file system in the `.dmg` file for simplicity. I also added device activation records to the `.dmg` file. There might have been some other changes that I had to do to boot the simulator but I didn't document them (sorry for that...).

**Comment by FrederickGeek8** [comment]
Is that a path I should go down at this juncture? I can't guarantee that the permissions are correct -- I `rsync`'ed with the archive option to a local folder, then did the same into a `.dmg` (and also tried creating it with Disk Utility from a folder). I feel like there are steps where permissions and the like could have been screwed up.

How did you go about merging the `boot` partition @devos50 ?

I also made an image of the device using `dd` since I thought that would be more complete and have better permissions. **I did it while the device was booted and running though and I think that's why I'm encountering issues trying to generate NAND from it**:
`ssh -oHostKeyAlgorithms=+ssh-rsa root@localhost "dd if=/dev/rdisk0 bs=1M" | dd of=image2.img status=progress`

`./generate_nand` doesn't appear to write the proper nand files. The `.img` is not mountable by default, but can be made mountable with [testdisk](https://www.cgsecurity.org/wiki/Recover_data_from_an_iPhone).
That does not solve the `./generate_nand` issue though. `testdisk` reports an invalid block0 DPME signature when I analyze it as an Apple partition map, so I guess my image is screwed up from the device being booted.
I also tried converting the `.img` to `.dmg` with diskutil and extracting the relevant partition with `dmg2img`. That too does not write the nand with `./generate_nand`.

I'm working on compiling OpeniBoot so I can see if I can pop a shell and dump the filesystem from there. Hopefully that get's me a more complete image. Do either of you know if I'll be able to install it without wiping out the data on the device? It's been a while since I worked with OpeniBoot and Google is terrible at finding the information I want.

Thanks for all the help :) It's fun getting to hack around iPhone OS once again.

**Comment by zoe-vb** [comment]
You can absolutely use openiboot without losing data, you don’t even need to flash it to use it, just use loadibec and then oibc to connect to it in shell mode

**Comment by FrederickGeek8** [comment]
I have some good news and some bad news.

Good news:
I was able to successfully boot an SSH Ramdisk with the `dd` command installed so that I could exfiltrate a disk image over SSH. I posted my fork of `msftguy/ssh-rd` [here](https://github.com/FrederickGeek8/ssh-rd).

Bad news:
While I was trying to boot a custom Ramdisk _manually_ (not using the fork I was able to fix up), I think I wiped out the drive on the device. I was able to mount it when I had my SSH Ramdisk without `dd` booted before my tinkering, however, now I am unable to see the partitions. I exfiltrated `disk0` anyways and opened it in a hex editor, and it's a much of 0xFFFFFFFF's. I'm so sad...

I'd be excited to hear if someone is able to pull something similar off. Meanwhile I'll be greaving the loss of my data... I'm super bummed.
